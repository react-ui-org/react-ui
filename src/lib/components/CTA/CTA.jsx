import PropTypes from 'prop-types';
import React from 'react';
import { withProviderContext } from '../../provider';
import styles from './CTA.scss';

export const CTA = (props) => {
  const {
    align,
    children,
  } = props;

  const alignClass = (value) => {
    if (value === 'top') {
      return styles.isAlignedToTop;
    }

    if (value === 'middle') {
      return styles.isAlignedToMiddle;
    }

    if (value === 'bottom') {
      return styles.isAlignedToBottom;
    }

    return styles.isAlignedToBaseline;
  };

  return (
    <div
      className={[
        styles.root,
        alignClass(align),
      ].join(' ')}
    >
      {children}
    </div>
  );
};

CTA.defaultProps = {
  align: 'baseline',
};

CTA.propTypes = {
  /**
   * Vertical alignment of CTA elements.
   */
  align: PropTypes.oneOf(['top', 'middle', 'bottom', 'baseline']),
  /**
   * Nested elements. Supported types are:
   * * `CTAStart`
   * * `CTACenter`
   * * `CTAEnd`
   */
  children: PropTypes.node.isRequired,
};

export const CTAWithContext = withProviderContext(CTA, 'CTA');

export default CTAWithContext;
