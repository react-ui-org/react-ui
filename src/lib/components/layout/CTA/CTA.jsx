import PropTypes from 'prop-types';
import React from 'react';
import styles from './CTA.scss';

export const CTA = (props) => {
  const {
    align,
    children,
  } = props;

  const alignClass = (value) => {
    if (value === 'top') {
      return styles.isAlignedToTop;
    }

    if (value === 'middle') {
      return styles.isAlignedToMiddle;
    }

    if (value === 'bottom') {
      return styles.isAlignedToBottom;
    }

    return styles.isAlignedToBaseline;
  };

  return (
    <div
      className={[
        styles.root,
        alignClass(align),
      ].join(' ')}
    >
      {children}
    </div>
  );
};

CTA.defaultProps = {
  align: 'baseline',
};

CTA.propTypes = {
  /**
   * Vertical alignment of CTA elements.
   */
  align: PropTypes.oneOf(['top', 'middle', 'bottom', 'baseline']),
  /**
   * Slot for individual CTA elements: CTAStart, CTACenter, and CTAEnd.
   */
  children: PropTypes.node.isRequired,
};

export default CTA;
