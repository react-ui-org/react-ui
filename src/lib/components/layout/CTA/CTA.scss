@import '../../../styles/tools/breakpoints';
@import '../../../styles/tools/offset';

$_cta-breakpoint: lg;
$_cta-spacing: offset(1);
$_cta-column-gap: offset(3);

.root {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  margin: calc(-1 * #{$_cta-spacing});

  @include breakpoint-up($_cta-breakpoint) {
    display: grid;
    grid-template: 'start center end' auto #{'/'} 1fr auto 1fr;
    grid-column-gap: $_cta-column-gap;
    margin: 0;
  }
}

.start,
.center,
.end {
  margin: $_cta-spacing;

  @include breakpoint-up($_cta-breakpoint) {
    margin: 0;
  }
}

.start {
  flex-basis: 100%;

  @include breakpoint-up($_cta-breakpoint) {
    grid-area: start;
    justify-self: start;
  }
}

.center {
  @include breakpoint-up($_cta-breakpoint) {
    grid-area: center;
    justify-self: center;
  }
}

.end {
  @include breakpoint-up($_cta-breakpoint) {
    grid-area: end;
    justify-self: end;
  }
}

.isAlignedToTop {
  align-items: start;
}

.isAlignedToMiddle {
  align-items: center;
}

.isAlignedToBottom {
  align-items: end;
}

.isAlignedToBaseline {
  align-items: baseline;
}
