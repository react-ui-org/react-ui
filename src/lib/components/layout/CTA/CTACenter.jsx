import PropTypes from 'prop-types';
import React from 'react';
import styles from './CTA.scss';

export const CTACenter = (props) => {
  const {
    children,
  } = props;

  return (
    <div className={styles.center}>
      {children}
    </div>
  );
};

CTACenter.propTypes = {
  /**
   * Content of the center element.
   */
  children: PropTypes.node.isRequired,
};

export default CTACenter;
