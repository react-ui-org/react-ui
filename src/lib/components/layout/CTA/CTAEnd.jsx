import PropTypes from 'prop-types';
import React from 'react';
import styles from './CenteredCTA.scss';

const CenteredCTAEnd = (props) => {
  const {
    children,
  } = props;

  return (
    <div className={styles.end}>
      {children}
    </div>
  );
};

CenteredCTAEnd.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};

export default CenteredCTAEnd;
