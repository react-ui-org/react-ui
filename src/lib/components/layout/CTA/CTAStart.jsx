import PropTypes from 'prop-types';
import React from 'react';
import styles from './CTA.scss';

export const CTAStart = (props) => {
  const {
    children,
  } = props;

  return (
    <div className={styles.start}>
      {children}
    </div>
  );
};

CTAStart.propTypes = {
  /**
   * Content of the start element.
   */
  children: PropTypes.node.isRequired,
};

export default CTAStart;
