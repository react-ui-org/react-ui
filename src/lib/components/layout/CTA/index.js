export { default as CTA } from './CTA';
export { default as CTACenter } from './CTACenter';
export { default as CTAEnd } from './CTAEnd';
export { default as CTAStart } from './CTAStart';
