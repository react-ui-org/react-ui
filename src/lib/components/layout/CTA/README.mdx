---
name: CTA
menu: 'Components: Layouts'
route: /components/layout/cta
---

# CTA

The responsive CTA layout holds an action at the center and two additional
elements on sides.

import {
  Playground,
  Props,
} from 'docz'
import { Placeholder } from '../../../../docs/_components/Placeholder/Placeholder'
import { CTA } from './CTA'
import { CTACenter } from './CTACenter'
import { CTAEnd } from './CTAEnd'
import { CTAStart } from './CTAStart'

## Basic Usage

To implement the CTA component, you need to import it first:

```js
import { CTA } from '@react-ui-org/react-ui';
```

And use it:

<Playground>
  <Placeholder bordered>
    <CTA>
      <CTAStart>
        <Placeholder bordered>
          Start
        </Placeholder>
      </CTAStart>
      <CTACenter>
        <Placeholder bordered>
          Center
        </Placeholder>
      </CTACenter>
      <CTAEnd>
        <Placeholder bordered>
          End
        </Placeholder>
      </CTAEnd>
    </CTA>
  </Placeholder>
</Playground>

See [API](#api) for all available options.

## General Guidelines

- The CTA (call to action) layout is designed for an **action placed at the
  center** of its parent and two optional side elements. It works perfectly
  especially in composition with a load-more button.

- Place rather **a few smaller elements** into the layout, so they fit the
  screen even at the smallest size. For complex layouts and a lot of actions,
  consider using the [Toolbar](/components/layout/toolbar) layout.

ðŸ‘‰ On screen sizes smaller than 66 em (the `lg`
[breakpoint](/foundation/breakpoints)), the start element expands over the full
width of its parent. Please resize your browser rather than the playground to
see this work.

## Vertical Alignment

There are four options of vertical alignment: baseline (default), top, middle,
and bottom.

<Playground>
  <Placeholder bordered>
    <CTA align="bottom">
      <CTAStart>
        <Placeholder bordered>
          Start
        </Placeholder>
      </CTAStart>
      <CTACenter>
        <Placeholder bordered>
          Center<br />
          Element
        </Placeholder>
      </CTACenter>
      <CTAEnd>
        <Placeholder bordered>
          End
        </Placeholder>
      </CTAEnd>
    </CTA>
  </Placeholder>
</Playground>

## API

<Props of={CTA} />

## CTAStart

The start element of the CTA layout.

<Props of={CTAStart} />

## CTACenter

The center element of the CTA layout.

<Props of={CTACenter} />

## CTAEnd

The end element of the CTA layout.

<Props of={CTAEnd} />

---

Next: [Center â†’](/components/layout/center)
