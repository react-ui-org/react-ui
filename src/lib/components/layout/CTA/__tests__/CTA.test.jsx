import React from 'react';
import { shallow } from 'enzyme';
import CTA from '../CTA';

describe('rendering', () => {
  it('renders correctly with a single child', () => {
    const tree = shallow((
      <CTA>
        <span>content</span>
      </CTA>
    ));

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with multiple children', () => {
    const tree = shallow((
      <CTA>
        <span>content 1</span>
        <span>content 2</span>
        <span>content 3</span>
      </CTA>
    ));

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with all props', () => {
    const tree = shallow((
      <CTA align="middle">
        <span>content</span>
      </CTA>
    ));

    expect(tree).toMatchSnapshot();
  });
});
