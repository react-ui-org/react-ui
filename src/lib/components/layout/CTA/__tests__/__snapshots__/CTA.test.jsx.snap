// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly with a single child 1`] = `
<div
  className="root isAlignedToBaseline"
>
  <span>
    content
  </span>
</div>
`;

exports[`rendering renders correctly with all props 1`] = `
<div
  className="root isAlignedToMiddle"
>
  <span>
    content
  </span>
</div>
`;

exports[`rendering renders correctly with multiple children 1`] = `
<div
  className="root isAlignedToBaseline"
>
  <span>
    content 1
  </span>
  <span>
    content 2
  </span>
  <span>
    content 3
  </span>
</div>
`;
